lib/backend/repository/repository.go

```diff
@@ -68,10 +83,66 @@ func (rb *RepositoryBackend) BuildACI(layerNumber int, layerID string, dockerURL
 	}
 }
 
-func (rb *RepositoryBackend) protocol() string {
-	if rb.insecure {
-		return "http://"
-	} else {
-		return "https://"
+func checkRegistryStatus(statusCode int, hdr http.Header, version string) (bool, error) {
+	switch statusCode {
+	case http.StatusOK, http.StatusUnauthorized:
+		ok := true
+		if version == "v2" {
+			// v2 API requires this check
+			ok = hdr.Get("Docker-Distribution-API-Version") == "registry/2.0"
+		}
+		return ok, nil
+	case http.StatusNotFound:
+		return false, nil
+	}
+
+	return false, fmt.Errorf("unexpected http code: %d", statusCode)
+}
+
+func (rb *RepositoryBackend) supportsRegistry(indexURL string, version string) (schema string, ok bool, err error) {
+	URLPath := version
+	if version == "v1" {
+		// the v1 API defines this URL to check if the registry's status
+		URLPath = "v1/_ping"
+	}
+	URLStr := path.Join(indexURL, URLPath)
+
+	fetch := func(schema string) (res *http.Response, err error) {
+		url := schema + "://" + URLStr
+		req, err := http.NewRequest("GET", url, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		rb.setBasicAuth(req)
+
+		client := &http.Client{}
+		res, err = client.Do(req)
```

Comment:
there's no `http.Do()`, there's `http.Get()` but it doesn't allow messing with the headers.